moda-contracts/src/Marketplace.sol

```diff
@@ -0,0 +1,236 @@
+
+    // Errors
+    error CannotBeZeroAddress();
+    error TreasuryFeeCannotBeZero();
+    error TokenAmountCannotBeZero();
+    error MaxCountCannotBeZero();
+    error StartCannotBeAfterEnd(uint256 startTime, uint256 endTime);
+    error InsufficientSupply(uint256 remainingSupply);
+    error MaxSupplyReached(uint256 maxSupplyPerWallet);
+    error SaleNotStarted(uint256 startTime, uint256 currentTime);
+    error SaleHasEnded(uint256 endTime, uint256 currentTime);
+    error ReleasesIsNotRegistered();
+
+    /**
+     * @dev Constructor
+     * @param treasury_ - The address of the organizations treasury
+     * @param treasuryFee_ - The percentage that will be transferred
+     * to the treasury on each sale. Based on a denominator of 10_000 e.g. 1000 = 10%
+     * @param usdc - The USDC token address
+     * @param catalog - The address of the catalog contract
+     */
+    constructor(address payable treasury_, uint256 treasuryFee_, address usdc, address catalog) {
+        if (usdc == address(0)) revert CannotBeZeroAddress();
+        if (treasuryFee_ == 0) revert TreasuryFeeCannotBeZero();
+        if (treasury_ == address(0)) revert CannotBeZeroAddress();
+        treasury = treasury_;
+        treasuryFee = treasuryFee_;
+        _usdc = usdc;
+        _catalog = catalog;
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    // External Functions
+
+    /**
+     * @inheritdoc IMarketplace
+     */
+    function createSale(
+        address payable beneficiary,
+        address releases,
+        uint256 tokenId,
+        uint256 amountTotal,
+        uint256 pricePerToken,
+        uint256 startAt,
+        uint256 endAt,
+        uint256 maxCountPerWallet
+    ) external {
+        if (IReleasesRegistration(_catalog).getReleasesOwner(releases) == address(0)) {
+            revert ReleasesIsNotRegistered();
+        }
+        if (beneficiary == address(0)) revert CannotBeZeroAddress();
+        if (amountTotal == 0) revert TokenAmountCannotBeZero();
+        if (endAt != 0 && startAt > endAt) {
+            revert StartCannotBeAfterEnd(startAt, endAt);
+        }
+        if (maxCountPerWallet == 0) revert MaxCountCannotBeZero();
+        IERC1155(releases).safeTransferFrom(_msgSender(), address(this), tokenId, amountTotal, "");
+
+        sales[_msgSender()].push(
+            Sale({
+                releaseOwner: _msgSender(),
+                beneficiary: beneficiary,
+                releases: releases,
+                tokenId: tokenId,
+                amountRemaining: amountTotal,
+                amountTotal: amountTotal,
+                pricePerToken: pricePerToken,
+                startAt: startAt,
+                endAt: endAt,
+                maxCountPerWallet: maxCountPerWallet
+            })
+        );
+
+        emit SaleCreated(_msgSender(), sales[_msgSender()].length - 1);
+    }
+
+    /**
+     * @inheritdoc IMarketplace
+     */
+    function purchase(
+        address seller,
```

Comment:
Let's say a manager creates a sale. A fan must now pass in the manager's address instead of the releaseOwner (Artist in most cases). `seller` should be `releaseOwner` so all sales are mapped to a single owner. This will be easier if I want to look for sales using deadmau5' address